runtime/onert/core/src/ir/train/TrainableGraph.cc

```diff
@@ -125,9 +126,61 @@ const ITrainableOperation &TrainableGraph::operation(OperationIndex index) const
   return dynamic_cast<const ITrainableOperation &>(_graph.operations().at(index));
 }
 
+void TrainableGraph::assertValidTopologicalOrder(const std::vector<ir::OperationIndex> &order,
+                                                 const std::string &order_type = "forward") const
+{
+  auto compare = [](const ir::OperationIndex &i, const ir::OperationIndex &j) -> bool {
+    return i.value() < j.value();
+  };
+
+  std::map<ir::OperationIndex, uint32_t, decltype(compare)> position(compare);
+  for (uint32_t p = 0; p < order.size(); ++p)
+  {
+    auto index = order[p];
+    if (position.count(index))
+      throw std::runtime_error{"Invalid " + order_type + " topological order: duplicate node @" +
+                               std::to_string(index.value())};
+
+    position[index] = p;
+  }
+
+  operations().iterate([&](const ir::OperationIndex &index, const ir::IOperation &op) {
+    if (position.count(index) == 0)
+      return;
+
+    uint32_t p = position[index];
+
+    for (const auto &output : op.getOutputs() | ir::Remove::DUPLICATED | ir::Remove::UNDEFINED)
+    {
+      const auto &operand = operands().at(output);
+      for (const auto &use : operand.getUses())
+      {
+        if (position.count(use) == 0)
+          continue;
+
+        uint32_t q = position[use];
+        if (p > q)
+          throw std::runtime_error{
+            "Invalid " + order_type + " topological order: inversion between @" +
+            std::to_string(index.value()) + " and @" + std::to_string(use.value())};
+      }
+    }
+  });
+}
+
+void TrainableGraph::assertValidBackwardTopologicalOrder(
+  std::vector<ir::OperationIndex> order) const
+{
+  std::reverse(order.begin(), order.end());
+  assertValidTopologicalOrder(order, "backward");
+}
```

Comment:
An order is a backward topological order if and only if its reverse is a forward topological order.

Reply:
This has been moved to

https://github.com/Samsung/ONE/blob/6f0884b43f2650c7178581fe245e5660625fb87f/runtime/onert/core/src/ir/train/TrainableGraph.cc#L132-L133